File: Dynamic_Programming/2478. Number of Beautiful Partitions.cpp
/*
2478. Number of Beautiful Partitions
*/


// TC:O(n*k) SC:O(n*k)
class Solution {
    using LL = long long;
    LL M = 1e9+7;
public:
    int beautifulPartitions(string s, int k, int ml) {
        int n = s.size();
        vector<vector<LL>> dp(k+1, vector<LL>(n+1));
        dp[0][0] = 1;

        if(k*ml > n || !isPrime(s[0]) || isPrime(s[n-1])) return 0;

        for(int i = 1; i <= k; i++){
            int sum = 0;
            for(int j = i*ml; j + (k-i)*ml <= n; j++){
                // 確認之前的斷點是否合法 
                if(can_partition(s, j-ml)) sum = (sum + dp[i-1][j-ml]) % M;
                // 確認當前是否合法可以整併到後續的分類去
                if(can_partition(s, j)) dp[i][j] = sum;
            }
        }

        return dp[k][n];

    }

    bool isPrime(char ch){
        return ch == '2' || ch == '3' || ch == '5' || ch == '7';
    }
    bool can_partition(string& s, int j){
        return j == 0 || j == s.size() || !isPrime(s[j-1]) && isPrime(s[j]);
    }
};

/*
長度至少 minLength, 要分k組 
開頭要是prime 結尾要是非prime
問說要多少拆分方法
n = 1000

*/

File: Dynamic_Programming/2478. Number of Beautiful Partitions.py
# TC:O(n*k) SC:O(n*k)
class Solution:
    def beautifulPartitions(self, s: str, k: int, minLength: int) -> int:
        '''
        1. 切成k份
        2. 每一份最少minLength
        3. 開頭prime (2,3,5,7) 結尾非prime(1,4,6,8,9)
        '''
        MOD = 10 ** 9 + 7
        n = len(s)
        if n < k * minLength:
            return 0
        dp = [[0] * (n + 1) for _ in range(k + 1)]
        dp[0][0] = 1

        primeSet = {'2', '3', '5', '7'}
        for t in range(1, k + 1):
            cumulative = 0
            # i 表示切分後的前綴長度，至少要有 t * minLength 個字符
            for i in range(t * minLength, n + 1):
                if s[i - minLength] in primeSet:
                    cumulative += dp[t - 1][i - minLength]
                dp[t][i] = cumulative if s[i - 1] not in primeSet else 0
        return dp[k][n] % MOD

        '''
        @cache
        def dfs(i: int, cnt: int) -> int:
            if i < 0 and cnt == 0:
                return 1
            if i < 0 or cnt == 0:
                return 0
            endDigit = int(s[i])
            if endDigit not in nonPrimeSet:
                return 0
            ans = 0
            for j in range(i - minLength + 1, -1, -1):
                startDigit = int(s[j])
                if startDigit in primeSet:
                    ans = (ans + dfs(j - 1, cnt - 1)) % MOD
            return ans
        
        return dfs(n - 1, k) % MOD
        '''
